ssl/priority_queue.c

```diff
@@ -0,0 +1,390 @@
+ * (ternary or quaternary) heap might elicit a performance advantage via better
+ * cache access patterns.
+ */
+
+struct pq_heap_s {
+    void *data;     /* User supplied data pointer */
+    size_t index;   /* Constant index in elements[] */
+};
+
+struct pq_elem_s {
+    size_t posn;    /* Current index in heap[] or link in free list */
+#ifndef NDEBUG
+    int used;       /* Debug flag indicating that this is in use */
+#endif
+};
+
+struct ossl_pqueue_st
+{
+    struct pq_heap_s *heap;
+    struct pq_elem_s *elements;
+    int (*compare)(const void *, const void *);
+    size_t htop;        /* Highest used heap element */
+    size_t hmax;        /* Allocated heap & element space */
+    size_t freelist;    /* Index into elements[], start of free element list */
+};
+
+/*
+ * The initial and maximum number of elements in the heap.
+ */
+static const size_t min_nodes = 8;
+static const size_t max_nodes =
+        SIZE_MAX / (sizeof(struct pq_heap_s) > sizeof(struct pq_elem_s)
+                    ? sizeof(struct pq_heap_s) : sizeof(struct pq_elem_s));
+
+#ifndef NDEBUG
+/* Some basic sanity checking of the data structure */
+# define ASSERT_USED(pq, idx)                                               \
+    OPENSSL_assert(pq->elements[pq->heap[idx].index].used);                 \
+    OPENSSL_assert(pq->elements[pq->heap[idx].index].posn == idx)
+# define ASSERT_ELEM_USED(pq, elem)                                         \
+    OPENSSL_assert(pq->elements[elem].used)
+#else
+# define ASSERT_USED(pq, idx)
+# define ASSERT_ELEM_USED(pq, elem)
+#endif
+
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth factor is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use an expansion factor of 8 / 5 = 1.6
+ */
+static ossl_inline int compute_pqueue_growth(size_t target, size_t current)
+{
+    int err = 0;
+
+    while (current < target) {
+        if (current >= max_nodes)
+            return 0;
+
+        current = safe_muldiv_size_t(current, 8, 5, &err);
+        if (err)
+            return 0;
+        if (current >= max_nodes)
+            current = max_nodes;
+    }
+    return current;
+}
+
+static ossl_inline void pqueue_swap_elem(OSSL_PQUEUE *pq, size_t i, size_t j)
+{
+    struct pq_heap_s *h = pq->heap, t_h;
+    struct pq_elem_s *e = pq->elements;
+
+    ASSERT_USED(pq, i);
+    ASSERT_USED(pq, j);
+
+    t_h = h[i];
+    h[i] = h[j];
+    h[j] = t_h;
+
+    e[h[i].index].posn = i;
+    e[h[j].index].posn = j;
+}
+
+static ossl_inline void pqueue_move_elem(OSSL_PQUEUE *pq, size_t from, size_t to)
+{
+    struct pq_heap_s *h = pq->heap;
+    struct pq_elem_s *e = pq->elements;
+
+    ASSERT_USED(pq, from);
+
+    h[to] = h[from];
+    e[h[to].index].posn = to;
+}
```

Comment:
Should the `from` element be cleared/removed?

Reply:
It doesn't have to be.  This is only called from the pop function which deals with the from side.